Sources/AsyncHTTPClient/AsyncAwait/Transaction+StateMachine.swift

```diff
@@ -0,0 +1,552 @@
+
+            case .finished:
+                // body was released after the response was consumed
+                return .none
+
+            case .executing(_, _, .buffering),
+                 .executing(_, _, .waitingForRemote),
+                 .executing(_, _, .finished):
+                // user is consuming the stream with an iterator
+                return .none
+            }
+        }
+
+        enum ConsumeAction {
+            case succeedContinuation(CheckedContinuation<ByteBuffer?, Error>, ByteBuffer?)
+            case failContinuation(CheckedContinuation<ByteBuffer?, Error>, Error)
+            case askExecutorForMore(HTTPRequestExecutor)
+        }
+
+        struct TriedToRegisteredASecondConsumer: Error {}
+
+        mutating func consumeNextResponsePart(
+            streamID: HTTPClientResponse.Body.IteratorStream.ID,
+            continuation: CheckedContinuation<ByteBuffer?, Error>
+        ) -> ConsumeAction {
+            switch self.state {
+            case .initialized,
+                 .queued,
+                 .executing(_, _, .initialized):
+                preconditionFailure("If we receive a response body, we must have received a head before")
+
+            case .executing(_, _, .finished(_, _)):
+                preconditionFailure("This is an invalid state at this point. We are waiting for the request stream to finish to succeed or response stream.")
+
+            case .executing(let context, let requestState, .waitingForStream(var buffer, next: .askExecutorForMore)):
+                if buffer.isEmpty {
+                    self.state = .executing(context, requestState, .waitingForRemote(streamID, continuation))
+                    return .askExecutorForMore(context.executor)
+                } else {
+                    let toReturn = buffer.removeFirst()
+                    self.state = .executing(context, requestState, .buffering(streamID, buffer, next: .askExecutorForMore))
+                    return .succeedContinuation(continuation, toReturn)
+                }
+
+            case .executing(_, _, .waitingForStream(_, next: .error(let error))):
+                self.state = .finished(error: error, streamID)
+                return .failContinuation(continuation, error)
+
+            case .executing(_, _, .waitingForStream(let buffer, next: .eof)) where buffer.isEmpty:
+                self.state = .finished(error: nil, streamID)
+                return .succeedContinuation(continuation, nil)
+
+            case .executing(let context, let requestState, .waitingForStream(var buffer, next: .eof)):
+                assert(!buffer.isEmpty)
+                let toReturn = buffer.removeFirst()
+                self.state = .executing(context, requestState, .buffering(streamID, buffer, next: .eof))
+                return .succeedContinuation(continuation, toReturn)
+
+            case .executing(let context, let requestState, .buffering(let streamID, var buffer, next: .askExecutorForMore)):
```

Comment:
We should check that the `streamID` in the state is the same as the `streamID` from the parameter.